API and implementation to quarantine service backends

aditighag opened the pull request:
The API will allow users to set backend states. The primary motivation is to quarantine unreachable backends so that such
backends won't be selected for load-balancing service traffic.
The backend states can be set in response to various events like Kubernetes events (e.g., terminating endpoints), or service APIs (e.g., quarantine unreachable backends). Backend states are persisted across cilium agent restart.

```
$ cilium service list
ID   Frontend          Service Type   Backend
4    10.96.95.246:80   ClusterIP      1 => 10.244.0.58:80 (active)
                                      2 => 10.244.1.59:80 (active)
                                      3 => 10.244.1.8:80 (active)

$ cilium service update --backends 10.244.0.58:80 --states quarantined
Updating backend states
Updated service with 1 backends

$ cilium service list
ID   Frontend          Service Type   Backend
4    10.96.95.246:80   ClusterIP      1 => 10.244.0.58:80 (quarantined)
                                      2 => 10.244.1.59:80 (active)
                                      3 => 10.244.1.8:80 (active)

$ cilium service update --backends 10.244.0.58:80 --states active
Updating backend states
Updated service with 1 backends

$ cilium service list
ID   Frontend          Service Type   Backend
4    10.96.95.246:80   ClusterIP      1 => 10.244.0.58:80 (active)
                                      2 => 10.244.1.59:80 (active)
                                      3 => 10.244.1.8:80 (active)

```
See commits description for more details.


```release-note
Support setting service backend states such as quarantine, maintenance so that these backends are not selected for load-balancing service traffic. 

```

aditighag committed:
pkg/loadbalancer: Introduce backend state

Backend state will indicate a backend's condition for load-balancing service
traffic. The state field will be used to determine whether a backend is
actively load-balancing service traffic, and will be plumbed/removed
from the BPF maps accordingly.

Signed-off-by: Aditi Ghag <aditi@cilium.io>

aditighag committed:
loadbalancer: Use Terminating backend state

Refactor the existing logic that reads and processes
backend's terminating condition for graceful termination
to use the `BackendState` introduced in the previous
commit.

Signed-off-by: Aditi Ghag <aditi@cilium.io>

borkmann commented on pkg/k8s/watchers/watcher.go:
```diff
@@ -644,13 +644,17 @@ func genCartesianProduct(
 
 			if backendPort := backend.Ports[string(fePortName)]; backendPort != nil && feFamilyIPv6 == ip.IsIPv6(parsedIP) {
+				backendState := loadbalancer.BackendStateActive
+				if backend.Terminating {
```
This `backend.Terminating` comes from k8s api, right?

Propagation to BPF map and restoration is still on TODO, right?

aditighag committed:
loadbalancer: Validate state transitions

Backend state transitions can be triggered from
either kubernetes events or service APIs.

Validate the transitions to ensure the system
is in a sane state when states are updated -
BackendStateActive -> BackendStateTerminating, BackendStateQuarantined, BackendStateMaintenance
BackendStateTerminating -> No valid state transition
BackendStateQuarantined -> BackendStateActive, BackendStateTerminating
BackendStateMaintenance -> BackendStateActive

Signed-off-by: Aditi Ghag <aditi@cilium.io>

aditighag committed:
api,cmd/loadbalancer: Extend backend model with state

Example output :

```
$ cilium service list
ID   Frontend           Service Type   Backend
1    172.20.0.1:443     ClusterIP      1 => 192.168.60.11:6443 (active)
2    172.20.0.10:53     ClusterIP      1 => 10.12.0.65:53 (active)
3    172.20.0.10:9153   ClusterIP      1 => 10.12.0.65:9153 (active)
4    [fd03::a]:53       ClusterIP      1 => [fd00::9cbf]:53 (active)
5    [fd03::a]:9153     ClusterIP      1 => [fd00::9cbf]:9153 (active)
```

Signed-off-by: Aditi Ghag <aditi@cilium.io>

aditighag commented on cilium/cmd/service_update.go:
```diff
@@ -50,6 +51,7 @@ func init() {
 	serviceUpdateCmd.Flags().BoolVarP(&k8sClusterInternal, "k8s-cluster-internal", "", false, "Set service as cluster-internal for externalTrafficPolicy=Local")
 	serviceUpdateCmd.Flags().StringVarP(&frontend, "frontend", "", "", "Frontend address")
 	serviceUpdateCmd.Flags().StringSliceVarP(&backends, "backends", "", []string{}, "Backend address or addresses (<IP:Port>)")
+	serviceUpdateCmd.Flags().StringSliceVarP(&backendStates, "backend-states", "", []string{}, "Backend state(s) as {active(default),terminating,quarantined}")
```
I've extended this list with the 4th state in local commits.

aditighag committed:
cmd, test: Print backend slot in bpf list output

The backend slot is set for every backend, and
is useful for debugging purposes.
Also, update the test output where we were
doing exact string match.

While at it, annotate the output so that it's clear
what the output values denote.

```
Example (before) -

$ sudo cilium bpf lb list
SERVICE ADDRESS    BACKEND ADDRESS
172.20.0.1:443     192.168.60.11:6443 (6)

(after) -

$ sudo cilium bpf lb list
SERVICE ADDRESS    BACKEND ADDRESS (REVNAT_ID) (SLOT)
172.20.0.1:443     192.168.60.11:6443 (6) (1)

```

Signed-off-by: Aditi Ghag <aditi@cilium.io>

aditighag committed:
lbmap: Sort backends before populating BPF maps

Previously, `UpsertServiceParams` struct maintained a map of backends
that are then copied into a slice before plumbing BPF maps. Map iteration
is non-deterministic which can lead to different ordering of backends. This
can change backend slots, and hence disrupt existing backends connections (e.g.,
UDP traffic in socket-based load-balancing.) IDs are monotonically increasing
so sorting should hopefully minimize the churn in the maps.

Signed-off-by: Aditi Ghag <aditi@cilium.io>

aditighag committed:
bpf,lb: Define flags in backend map value

Use the padding bits as flags. This is a
preparatory commit for the subsequent commits to be
able to encode backend state in the flags field which
can be restored after agent restart.

Signed-off-by: Aditi Ghag <aditi@cilium.io>

aditighag committed:
cmd/loadbalancer: Add API to update service backend states

The API will allow users to update backend states. The primary
motivation is to quarantine unreachable backends so that such
backends won't be selected for load-balancing service traffic.
Internally, the agent iterate over the services map, and update all
the services that select the updated backends. States transitions
are internally validated before updating the states.

See dump below to how to set "quarantine" state for backends.
Quarantined backends when become reachable can be set back to
the active state.

```

$ cilium service list
ID   Frontend          Service Type   Backend
4    10.96.95.246:80   ClusterIP      1 => 10.244.0.58:80 (active)
                                      2 => 10.244.1.59:80 (active)
                                      3 => 10.244.1.8:80 (active)

$ cilium service update --backends 10.244.0.58:80 --states quarantined
Updating backend states
Updated service with 1 backends

$ cilium service list
ID   Frontend          Service Type   Backend
4    10.96.95.246:80   ClusterIP      1 => 10.244.0.58:80 (quarantined)
                                      2 => 10.244.1.59:80 (active)
                                      3 => 10.244.1.8:80 (active)

$ cilium service update --backends 10.244.0.58:80 --states active
Updating backend states
Updated service with 1 backends

$ cilium service list
ID   Frontend          Service Type   Backend
4    10.96.95.246:80   ClusterIP      1 => 10.244.0.58:80 (active)
                                      2 => 10.244.1.59:80 (active)
                                      3 => 10.244.1.8:80 (active)

$ cilium service update --backends 10.244.0.58:80 --states terminating
Updating backend states
Updated service with 1 backends

e$ cilium service list
ID   Frontend          Service Type   Backend
4    10.96.95.246:80   ClusterIP      1 => 10.244.0.58:80 (terminating)
                                      2 => 10.244.1.59:80 (active)
                                      3 => 10.244.1.8:80 (active)

$ cilium service update --backends 10.244.0.58:80 --states active
Updating backend states
Error: Cannot add/update service: [PUT /service/{id}][501] putServiceIdUpdateBackendFailure  invalid state transition for backend[10.244.0.58:80] (terminating) -> (active) &{BackendAddresses:[0xc0000f8990] Flags:0xc0001004b0 FrontendAddress:0xc00012f9c0 ID:0}

```

Signed-off-by: Aditi Ghag <aditi@cilium.io>

aditighag committed:
lb: Encode backend state in backend map value

The `flags` field in the backend map value will be
used to encode the backend's state. This helps in
restoring backend states after agent restart.

Signed-off-by: Aditi Ghag <aditi@cilium.io>

aditighag committed:
pkg/service: Support updating backend states for multiple services

This commit introduces changes to update multiple services that
have backends whose states are being updated. This will be used by
a new service API to update backend states (e.g., quarantine backends),
whereby backend states for all the associated services are updated in a
consistent manner with a single call.

Added unit tests to validate the functionality.

Signed-off-by: Aditi Ghag <aditi@cilium.io>

aditighag committed:
lb: Restore backend state after agent restart

This commit builds up on previous commits where
we restore non-active backends after agent restart.
The non-active backend states such as `Quarantined` will
not be synced with Kubernetes events as these are set with
the Cilium service API. As a result, we need to persist backend
state in BPF maps, and restore it after agent restart.

When the agent is restarted, we first restore backends and services
from BPF maps. The persisted backends are marked as restored from
datapath. Later, once the agent's sync with Kubernetes API server
is complete, we reconcile the state restored from BPF maps with the service
and endpoint related events received from Kubernetes. The backend flag that
tracks restored backends from datapath is used to restore backend state
during this phase.

Testing -

Before agent restart -

```
$ cilium service list
ID   Frontend          Service Type   Backend
4    10.96.95.246:80   ClusterIP      1 => 10.244.0.58:80 (active)
                                      2 => 10.244.1.59:80 (active)
                                      3 => 10.244.1.8:80 (active)

$ sudo bpftool map dump pinned /sys/fs/bpf/tc/globals/cilium_lb4_backends_v2
key: 03 00 00 00  value: 0a f4 01 08 00 50 00 00
key: 02 00 00 00  value: 0a f4 01 3b 00 50 00 00
key: 01 00 00 00  value: 0a f4 00 3a 00 50 00 00

$ cilium service update --backends 10.244.0.58:80 --states quarantined
Updating backend states
Updated service with 1 backends

$ sudo bpftool map dump pinned /sys/fs/bpf/tc/globals/cilium_lb4_backends_v2
key: 03 00 00 00  value: 0a f4 01 08 00 50 00 00
key: 02 00 00 00  value: 0a f4 01 3b 00 50 00 00
key: 01 00 00 00  value: 0a f4 00 3a 00 50 00 02
Found 3 elements

$ sudo systemctl restart cilium

```

After restart -

```

$ sudo bpftool map dump pinned /sys/fs/bpf/tc/globals/cilium_lb4_backends_v2
key: 03 00 00 00  value: 0a f4 01 08 00 50 00 00
key: 02 00 00 00  value: 0a f4 01 3b 00 50 00 00
key: 01 00 00 00  value: 0a f4 00 3a 00 50 00 02
Found 3 elements

$ cilium service list
ID   Frontend          Service Type   Backend
4    10.96.95.246:80   ClusterIP      1 => 10.244.0.58:80 (quarantined)
                                      2 => 10.244.1.59:80 (active)
                                      3 => 10.244.1.8:80 (active)

```

Signed-off-by: Aditi Ghag <aditi@cilium.io>

aditighag committed:
lb: Persist updated backend states in BPF maps

Backend states can change until the backends are deleted.
The new states need to be updated in the BPF backend map
so that the last updated state can be restored after agent restart.

```
$ cilium service list
ID   Frontend          Service Type   Backend
4    10.96.95.246:80   ClusterIP      1 => 10.244.0.58:80 (active)
                                      2 => 10.244.1.59:80 (active)
                                      3 => 10.244.1.8:80 (active)

$ cilium service update --backends 10.244.0.58:80 --states quarantined
Updating backend states
Updated service with 1 backends

$ sudo bpftool map dump pinned /sys/fs/bpf/tc/globals/cilium_lb4_backends_v2
key: 03 00 00 00  value: 0a f4 01 08 00 50 00 00
key: 02 00 00 00  value: 0a f4 01 3b 00 50 00 00
key: 01 00 00 00  value: 0a f4 00 3a 00 50 00 02

$ cilium service update --backends 10.244.0.58:80 --states active
Updating backend states
Updated service with 1 backends

$ cilium service list
ID   Frontend          Service Type   Backend
4    10.96.95.246:80   ClusterIP      1 => 10.244.0.58:80 (active)
                                      2 => 10.244.1.59:80 (active)
                                      3 => 10.244.1.8:80 (active)

$ sudo bpftool map dump pinned /sys/fs/bpf/tc/globals/cilium_lb4_backends_v2
key: 03 00 00 00  value: 0a f4 01 08 00 50 00 00
key: 02 00 00 00  value: 0a f4 01 3b 00 50 00 00
key: 01 00 00 00  value: 0a f4 00 3a 00 50 00 00
Found 3 elements

```

Signed-off-by: Aditi Ghag <aditi@cilium.io>

aditighag committed:
test: Add runtime test to validate backend state updates

Validates service API to update backend states reflects
updated state in services list. Also, tests restore workflow
after agent restart.

Signed-off-by: Aditi Ghag <aditi@cilium.io>

ldelossa commented on pkg/loadbalancer/loadbalancer.go:
```diff
@@ -182,6 +217,9 @@ type BackendID uint32
 type ID uint32
 
+// BackendState is the state of a backend for load-balancing service traffic.
+type BackendState uint8
```
Adding a `String()` methid to this type which prints out a string repr given the receiver's integeger could be helpful for logging/debugging purposes.

ldelossa commented on pkg/service/service.go:
```diff
@@ -1053,6 +1053,13 @@ func (s *Service) updateBackendsCacheLocked(svc *svcInfo, backends []lb.Backend)
 			svc.backendByHash[hash] = &backends[i]
 		} else {
 			backends[i].ID = b.ID
+			if b.State != backends[i].State {
```
What is the ultimate reprocussion for a backend entering the business logic with an invalid transition? 

Could this situation be avoided completely during runtime by sanitizing before the backends are even written to APIs which trigger events?

aditighag replied:
>Could this situation be avoided completely during runtime by sanitizing before the backends are even written to APIs which trigger events?

The backend states are validated when they are set at the API handler, but the transitions can only be validated where we have access to the previously saved backend state (which is done in this function).

ldelossa commented on pkg/loadbalancer/loadbalancer.go:
```diff
@@ -315,6 +315,36 @@ func IsValidStateTransition(old, new BackendState) bool {
 	return true
 }
 
+func GetBackendState(state string) (BackendState, error) {
```
nit: this maybe a little nicer implemented on the `BackendState` type itself.

aditighag replied:
This can't be implemented with `BackendState` as the receiver as it takes string type as an argument. But I agree with your suggestion below for `GetBackendStateString`.

ldelossa replied:
It would look like 

var t BackendState
t.FromString("active") 

If you wanted to fill in the var's mem address with the int representing the incoming state.

ldelossa commented on pkg/loadbalancer/loadbalancer.go:
```diff
@@ -315,6 +315,36 @@ func IsValidStateTransition(old, new BackendState) bool {
+	}
+}
+
+func GetBackendStateString(state BackendState) (string, error) {
```
this maybe a little nicer implemented on the `BackendState` type itself. (Same as GetBackendState)

ldelossa commented on pkg/maps/lbmap/lbmap.go:
```diff
@@ -72,9 +73,15 @@ type UpsertServiceParams struct {
+// will not be considered for load-balancing traffic. The non-active backends
+// are, however, populated in the service map so that they can be restored
+// upon agent restart.
 func (lbmap *LBBPFMap) UpsertService(p *UpsertServiceParams) error {
```
> The service's non-active backends are appended to the active backends list,
Why is this the case? 

Seems a bit odd to have non-active backends in the active backend list. 

If this is just for logging/cli dumping is it possible to dump the `NonActiveBackends` and omit the need to add them to the `ActiveBackends` set?

This would also reduce the code involved in separating the active/nonactive backends in service.go

aditighag replied:
>Seems a bit odd to have non-active backends in the active backend list.

That's not the case. They are initially segregated so that we can set the service's backend count to only the active backends. But they are required to be persisted to service maps for restore purposes as the note in the function description indicates. The service's backend count is set in the datapath as part of the master key (I'll update the comment with this detail). So in the datapath, first lookup is done to check if the service has any backends. If it does, then a follow-up lookup is made to select a backend slot.

ldelossa replied:
Got it

ldelossa commented on pkg/service/service.go:
```diff
@@ -1142,3 +1240,23 @@ func isWildcardAddr(frontend lb.L3n4AddrID) bool {
 	return net.IPv4zero.Equal(frontend.IP)
 }
+
+func segregateBackends(backends []lb.Backend) (map[string]lb.BackendID, []lb.BackendID) {
```
worth a comment on this function.

aditighag replied:
This function isn't exported. The function has detailed comments to explain what it does. But to make it more explicit, I used named return parameters.

ldelossa commented on bpf/lib/common.h:
```diff
@@ -748,7 +748,7 @@ struct lb6_backend {
 	__be16 port;
 	__u8 proto;
-	__u8 pad;
+	__u8 flags;
```
Should this be called `state`. 

IMO 'flags" makes me think this field is used for more then one config/purpose. But I think the usage just evaluates this field as the integer representing the backend state. Maybe better to just call it what it is.

aditighag replied:
Flags can be extended in future for other purposes so I'll leave it as "flags".

ldelossa reviewed (changes requested):
Nice feature. 

Couple of questions about implementation.

aditighag reviewed (commented):
Thanks for the review. I've addressed some of the review comments, replied inline to others.

sayboras commented on pkg/loadbalancer/loadbalancer.go:
```diff
@@ -358,6 +398,7 @@ func NewL3n4AddrFromModel(base *models.FrontendAddress) (*L3n4Addr, error) {
 }
 
 // NewBackend creates the Backend struct instance from given params.
+// The default state for the returned Backend is BackendStateActive.
```
nit: we can explicitly set State: BackendStateActive for returned b as well.

sayboras commented on pkg/loadbalancer/loadbalancer.go:
```diff
@@ -256,6 +296,28 @@ func (s *SVC) GetModel() *models.Service {
 	}
 }
 
+func IsValidStateTransition(old, new BackendState) bool {
```
nit: unlikely new state param is BackendStateInvalid, just to be exhaustive

sayboras commented on api/v1/openapi.yaml:
```diff
@@ -643,6 +643,11 @@ paths:
           x-go-name: Failure
           schema:
             "$ref": "#/definitions/Error"
+        '501':
```
AFAIK, 501 is used if the server didn't recognize method (e.g. PUT in this case). 

can you explain why 501 error code is used here ? Or due to nature of internal API, we are following our own convention (i.e. 500 for creating service error, and 501 for updating state error).

aditighag replied:
>we are following our own convention (i.e. 500 for creating service error, and 501 for updating state error).

Yes, this is in line with the existing update code.

sayboras commented on cilium/cmd/service_update.go:
```diff
@@ -50,6 +52,8 @@ func init() {
 	serviceUpdateCmd.Flags().StringVarP(&frontend, "frontend", "", "", "Frontend address")
 	serviceUpdateCmd.Flags().StringSliceVarP(&backends, "backends", "", []string{}, "Backend address or addresses (<IP:Port>)")
+	serviceUpdateCmd.Flags().StringSliceVarP(&backendStates, "states", "", []string{}, "Backend state(s) as {active(default),terminating,quarantined,maintenance}")
+	serviceUpdateCmd.Flags().BoolVarP(&updateServices, "update-services", "", false, "Update all services selecting the backends with their given states (id and frontend are ignored)")
```
while I understand why this flag is added, it's a little confusing to me :(.

Just curious if you have considered to have a sub command (e.g. cilium service update state --backends 1.1.1.1:80 --stage quarantined)

Also, is there any case that --states flag can be used independently with --update-services flag ?

aditighag replied:
>Also, is there any case that --states flag can be used independently with --update-services flag ?

There is no use case right now. 

Regarding sub-command, we internally discussed to add a new flag. The `update-services` flag indicates that multiple services can potentially be updated.

borkmann replied:
Right now you have `cilium service update --backends 10.244.0.58:80 --states quarantined --updates-services=true`.

The last `--updates-services=true` is very odd from a user pov, why can't we imply this when no frontend parameter was passed so that a `cilium service update --backends 10.244.0.58:80 --states quarantined` is sufficient?

aditighag replied:
@borkmann I had brought this up last month, and we decided to not rely on implicit assumptions, rather make the API intent more clear by adding a flag. I'll cc you to that thread.

aditighag replied:
Removed the flag as discussed offline.

sayboras commented on test/helpers/cilium.go:
```diff
@@ -805,7 +805,8 @@ func (s *SSHMeta) ServiceIsSynced(id int) (bool, error) {
 
 		for _, addr := range lb {
-			if addr == target {
+			// if addr == target {
```
maybe just remove this line

sayboras reviewed (commented):
Couple of comments as per below.

This feature is working very well with global service in clustermesh as well, I have tested one case in which the remote services are marked as quarantined, requests are hitting only local endpoints as expected.   

One more question about pod recycling, if the pod is recycled, will the previous configuraiton (e.g. quarantined) be lost as backend IP will be something else ? :thinking:


<details>

```
# Cluster 2: marking remote backend from cluster 1 as quarantined

root@kind-2-control-plane:/home/cilium# cilium service list
ID   Frontend             Service Type   Backend                              
1    10.96.0.10:53        ClusterIP      1 => 10.244.1.197:53 (active)        
                                         2 => 10.244.1.24:53 (active)         
2    10.96.0.10:9153      ClusterIP      1 => 10.244.1.197:9153 (active)      
                                         2 => 10.244.1.24:9153 (active)       
3    10.96.0.1:443        ClusterIP      1 => 172.18.0.6:6443 (active)        
4    10.96.191.183:2379   ClusterIP      1 => 10.244.1.130:2379 (active)      
5    10.96.22.51:80       ClusterIP      1 => 10.244.0.55:80 (active)         
                                         2 => 10.244.0.46:80 (quarantined)    
                                         3 => 10.244.0.19:80 (active)         
                                         4 => 10.244.1.223:80 (quarantined)   

$ ./local/clustermesh_check.sh
Checking in cluster 1
{"Galaxy": "Alderaan", "Cluster": "Cluster-2"}
{"Galaxy": "Alderaan", "Cluster": "Cluster-2"}
{"Galaxy": "Alderaan", "Cluster": "Cluster-2"}
{"Galaxy": "Alderaan", "Cluster": "Cluster-1"}
{"Galaxy": "Alderaan", "Cluster": "Cluster-1"}
{"Galaxy": "Alderaan", "Cluster": "Cluster-2"}
{"Galaxy": "Alderaan", "Cluster": "Cluster-2"}
{"Galaxy": "Alderaan", "Cluster": "Cluster-1"}
{"Galaxy": "Alderaan", "Cluster": "Cluster-1"}
{"Galaxy": "Alderaan", "Cluster": "Cluster-2"}
Checking in cluster 2
{"Galaxy": "Alderaan", "Cluster": "Cluster-2"}
{"Galaxy": "Alderaan", "Cluster": "Cluster-2"}
{"Galaxy": "Alderaan", "Cluster": "Cluster-2"}
{"Galaxy": "Alderaan", "Cluster": "Cluster-2"}
{"Galaxy": "Alderaan", "Cluster": "Cluster-2"}
{"Galaxy": "Alderaan", "Cluster": "Cluster-2"}
{"Galaxy": "Alderaan", "Cluster": "Cluster-2"}
{"Galaxy": "Alderaan", "Cluster": "Cluster-2"}
{"Galaxy": "Alderaan", "Cluster": "Cluster-2"}
{"Galaxy": "Alderaan", "Cluster": "Cluster-2"}
```

</details>

sayboras reviewed (approved):
LGTM :100:

aditighag committed:
lb: Persist non-active backends for restore

Commit b79a4a5370 skipped adding terminating backends to the service
BPF maps so that they are not selected for load-balancing service traffic.
With the addition of new states for backends (e.g., quarantined, maintenance),
we need to persist these non-active backends so that they can be restored
after agent restart.

To that end, this commit makes following changes :

- Maintain a separate list of non-active backends so that
  they can be added to the services map. Such backends will
  be appended to the active backends list. The main reason to
  append them to the end of active backends list is to minimize
  disruption caused due to updates of these non-active backends.
- A service's backends count in the datapath is set to include
  only the active backends so that non-active backends won't
  be selected for load-balancing new service traffic.
- As we maintain all service's backend (active and non-active),
  update the logic that keeps track of all the previously
  stored backends.
- Update unit tests accordingly.

Testing -

```

$ cilium service list
ID   Frontend          Service Type   Backend
4    10.96.95.246:80   ClusterIP      1 => 10.244.0.58:80 (active)
                                      2 => 10.244.1.59:80 (active)
                                      3 => 10.244.1.8:80 (active)

$ sudo cilium bpf lb list
SERVICE ADDRESS   BACKEND ADDRESS (REVNAT_ID) (SLOT)
10.96.95.246:80   0.0.0.0:0 (4) (0) [ClusterIP, non-routable]
                  10.244.1.8:80 (4) (3)
                  10.244.1.59:80 (4) (2)
                  10.244.0.58:80 (4) (1)

```
Change backend `10.244.0.58` state to terminating - notice the ordered list
of backends - active backends followed by non-active.

$ cilium service list
ID   Frontend          Service Type   Backend
4    10.96.95.246:80   ClusterIP      1 => 10.244.0.58:80 (terminating)
                                      2 => 10.244.1.59:80 (active)
                                      3 => 10.244.1.8:80 (active)

$ sudo cilium bpf lb list
SERVICE ADDRESS   BACKEND ADDRESS (REVNAT_ID) (SLOT)
10.96.95.246:80   0.0.0.0:0 (4) (0) [ClusterIP, non-routable]
                  10.244.0.58:80 (4) (3)
                  10.244.1.8:80 (4) (2)
                  10.244.1.59:80 (4) (1)

```

Suggested-by : Daniel Borkmann <daniel@cilium.io>
Signed-off-by: Aditi Ghag <aditi@cilium.io>

borkmann commented on cilium/cmd/service_update.go:
```diff
@@ -82,10 +84,23 @@ func boolToInt(set bool) int {
+		// When service ID is 0 and frontend is not specified, the intended use
+		// of the API is to update backend states for service(s) selecting those
+		// backends.
+		if len(backends) == 0 || len(backendStates) == 0 {
```
small nit: `len(backends) == 0` can never be hit given the `case` above?

borkmann reviewed (approved):
🚀
